### ISSUE TITLE
chore(typing): wire runtime shape/dtype checking via jaxtyping + beartype

### ISSUE THREAD

**Issue by antonioterpin** [antonioterpin/pinet#99]
## Motivation

Once [#97](https://github.com/antonioterpin/pinet/issues/97) lands, our function signatures will encode dtype and symbolic shapes (`Float[Array, "B m n"]`). The annotations alone give us static-checker leverage but no runtime guarantee — a caller can still pass an `(m, n)` array instead of `(B, m, n)` and only discover the mismatch deep inside a traced function, with a misleading error.

jaxtyping pairs cleanly with [beartype](https://beartype.readthedocs.io/) (or typeguard) via `jaxtyping.install_import_hook` / `@jaxtyping.jaxtyped`. The wrapper runs at **trace time**, so checks cost nothing during execution of the compiled artifact — we catch shape bugs once, at the traced call, and then JIT runs unimpeded.

We currently enforce these invariants ad-hoc with asserts, e.g.:

- [src/pinet/constraints/affine_equality.py:65-90](src/pinet/constraints/affine_equality.py#L65-L90) — `setup()` asserts `ndim == 3`, `b.shape[2] == 1`, batch compatibility.
- [src/pinet/dataclasses.py:94-151](src/pinet/dataclasses.py#L94-L151) — `BoxConstraintSpecification._validate_*` manually walks `ndim`, shapes, mask dtype.
- [src/pinet/dataclasses.py:211-285](src/pinet/dataclasses.py#L211-L285) — `SocConstraintSpecification.validate` re-implements the same pattern.
- [src/pinet/equilibration.py:30-33](src/pinet/equilibration.py#L30-L33) — `ruiz_equilibration` asserts the matrix is 2D.

All of these disappear once the annotation itself is the contract.

## Proposal

1. Add `beartype` (or `typeguard` — beartype is faster and more actively maintained) as a dev/runtime dependency.
2. In `src/pinet/__init__.py`, install the jaxtyping import hook scoped to our package:
   ```python
   from jaxtyping import install_import_hook
   with install_import_hook("pinet", "beartype.beartype"):
       from . import project, constraints, solver, equilibration, dataclasses  # noqa: F401
   ```
   This wraps every function/class defined inside `pinet/` automatically; user code is untouched.
3. Gate it behind an env var (e.g. `PINET_RUNTIME_CHECK=1`) so production/training runs can opt out if tracing overhead ever matters. Default to **on** for tests and dev — the whole point is to catch shape bugs during development.
4. Delete the hand-rolled `validate()` methods once they become strictly a subset of what jaxtyping enforces. Keep the *semantic* checks that jaxtyping cannot express (e.g. "lb <= ub elementwise" in [dataclasses.py:134](src/pinet/dataclasses.py#L134), "mask_t has exactly one True" in [dataclasses.py:242-246](src/pinet/dataclasses.py#L242-L246)).

## Open questions

- Should the check run in tests unconditionally? Probably yes — tests are where we most benefit from the immediate error.
- Does `install_import_hook` play nicely with our `jax.custom_vjp` and the `partial(...)` patterns in [project.py](src/pinet/project.py)? Worth prototyping on a single file before rolling out.
- Pick one of `beartype` vs `typeguard`. Recommendation: beartype — actively maintained, explicitly supported by jaxtyping, zero-config.

## Definition of done

- [ ] jaxtyping hook installed in `pinet/__init__.py` behind an env flag.
- [ ] `uv run pytest` runs with the hook enabled and still passes.
- [ ] Hand-rolled `validate()` methods pruned to only the constraints jaxtyping cannot express.
- [ ] A single failing call (e.g. passing a 2D array where `Float[Array, "B m n"]` is expected) produces a readable error at the call site, not deep in a traced function.

Depends on [#97](https://github.com/antonioterpin/pinet/issues/97).

**Comment by antonioterpin** [comment]
Landed via #113 (runtime typecheck infra) + #114 (flipped default to on, fixed 298 tests).
